lib/middleware/resource.js

```diff
@@ -1,47 +1,18 @@
 const debug = require('debug')('hydra-box:resource')
-const clownface = require('clownface')
-const ns = require('../namespaces')
-const { fromStream } = require('rdf-dataset-ext')
-const rdf = { ...require('@rdfjs/data-model'), ...require('@rdfjs/dataset') }
-const TermSet = require('@rdfjs/term-set')
 
-async function loadResource (store, term) {
-  const dataset = await fromStream(rdf.dataset(), store.match(null, null, null, term))
-
-  if (dataset.size === 0) {
-    return null
-  }
-
-  const types = new TermSet(clownface({ dataset, term }).out(ns.rdf.type).terms)
-
-  return {
-    term,
-    dataset,
-    types
-  }
-}
-
-async function findClassResource (store, term) {
-  return loadResource(store, term)
-}
-
-async function findPropertyResource (store, propertyTerm) {
-  const dataset = await fromStream(rdf.dataset(), store.match(null, null, propertyTerm, null))
-
-  if (dataset.size === 0) {
-    return null
-  }
+function factory ({ loader }) {
+  return async (req, res, next) => {
+    let resources = await loader.byClass(req.hydra.term)
 
-  const term = [...dataset][0].subject
+    if (resources.length === 0) {
+      resources = await loader.byProperty(req.hydra.term)
+    }
 
-  return loadResource(store, term)
-}
+    if (resources.length !== 1) {
+      return next(new Error(`no unique resource found for: <${req.hydra.term.value}>`))
+    }
 
-function factory (store) {
-  return async (req, res, next) => {
-    req.hydra.resource =
-      await findClassResource(store, req.hydra.term) ||
-      await findPropertyResource(store, req.hydra.term)
+    req.hydra.resource = resources[0]
```

Comment:
Why fail when there are `>1` hits? You plan reworking that in a next step?

Specifically, there might be a case when `term` is an object of multiple relations, found by `byProperty`. The request should still succeed if we're able to find a single operation to handle it.

Reply:
This will be covered in the next step with the operation finder.